src/platform_impl/macos/mod.rs

```diff
@@ -1 +1,526 @@
+mod accelerator;
 
+use std::{cell::RefCell, collections::HashMap, rc::Rc};
+
+use cocoa::{
+    appkit::{NSApp, NSApplication, NSEventModifierFlags, NSMenu, NSMenuItem},
+    base::{id, nil, selector, NO, YES},
+    foundation::{NSAutoreleasePool, NSInteger, NSString},
+};
+use objc::runtime::{Object, Sel};
+
+use crate::{
+    accelerator::Accelerator,
+    MenuEntryType,
+    predefined::PredfinedMenuItemType,
+    util::{AddOp, Counter},
+};
+
+static COUNTER: Counter = Counter::new();
+
+/// A generic child in a menu
+///
+/// Be careful when cloning this item and treat it as read-only
+#[derive(Debug, Default)]
+#[allow(dead_code)]
+struct MenuChild {
+    // shared fields between submenus and menu items
+    type_: MenuEntryType,
+    text: String,
+    enabled: bool,
+
+    ns_menuitems: HashMap<Menu, id>,
+
+    // menu item fields
+    id: u32,
+    accelerator: Option<Accelerator>,
+
+    // predefined menu item fields
+    is_predefined: bool,
+    predefined_item_type: PredfinedMenuItemType,
+
+    // check menu item fields
+    checked: bool,
+
+    // submenu fields
+    submenu: Option<Menu>,
+    children: Option<Vec<Rc<RefCell<MenuChild>>>>,
+}
+
+impl MenuChild {
+    pub fn id(&self) -> u32 {
+        self.id
+    }
+
+    pub fn text(&self) -> String {
+        self.text.clone()
+    }
+
+    pub fn set_text(&mut self, text: &str) {
+        self.text = text.to_string().replace("&", "");
+        unsafe {
+            let title = NSString::alloc(nil).init_str(&self.text);
+            for (_, &ns_item) in &self.ns_menuitems {
+                let () = msg_send![ns_item, setTitle: title];
+                let ns_submenu: *mut Object = msg_send![ns_item, submenu];
+                if ns_submenu != nil {
+                    let () = msg_send![ns_submenu, setTitle: title];
+                }
+            }
+        }
+    }
+
+    pub fn is_enabled(&self) -> bool {
+        self.enabled
+    }
+
+    pub fn set_enabled(&mut self, enabled: bool) {
+        self.enabled = enabled;
+        unsafe {
+            for (_, &ns_item) in &self.ns_menuitems {
+                let () = msg_send![ns_item, setEnabled: if enabled { YES } else { NO }];
+            }
+        }
+    }
+
+    pub fn is_checked(&self) -> bool {
+        self.checked
+    }
+
+    pub fn set_checked(&mut self, checked: bool) {
+        self.checked = checked;
+        unsafe {
+            for (_, &ns_item) in &self.ns_menuitems {
+                let () = msg_send![ns_item, setState: if checked { 1_isize } else { 0_isize }];
+            }
+        }
+    }
+}
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash)]
+pub struct Menu {
+    pub ns_menu: id,
```

Comment:
why is `ns_menu` public?

Reply:
Good point, I don't think it needs to be.